### ISSUE TITLE
[@types/react-redux] [enhancement] Connected component type inference

### ISSUE THREAD

**Issue by Voronar** [DefinitelyTyped/DefinitelyTyped#31227]
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @charrondev, @tkqubo, @kenzierocks, @clayne11, @tansongyang, @NicholasBoll, @mDibyo, @pdeva, @Kallikrein, @val1984, @jrakotoharisoa, @apapirovski, @surgeboris

When we work with TypeScript we have to manually declare prop types injected to connected component. 
In [this article](https://dev.to/voronar/practical-typescript-react--redux-4enh) I described how to do this work automatically with type inference.
Do you think about what? What about providing this utility type to main typings?

**Comment by Voronar** [comment]
Hello @Kallikrein !
I don't see the problem. Code you have write above is correctly work with `TypeOfConnect` utility:
```typescript
const myStateToProps = (_: any, ownProps: {baz: string}): {foo: string} => ({
  foo: ownProps.baz,
});

const enhancer = connect(myStateToProps);

type Props = {
  bar: string;
}
  & TypeOfConnect<typeof enhancer>
;

const MyInnerComponent: React.StatelessComponent<Props> = props => (
  <div>
    <div className="foo">{props.foo}</div>
    <div className="bar">{props.bar}</div>
  </div>
);

const MyOuterComponent = enhancer(MyInnerComponent);

<MyOuterComponent bar="" baz="">1</MyOuterComponent>;
```

**Comment by surgeboris** [comment]
@Voronar Thank you very much for the article, I like the approach and I think that this utility type  might be really usefull.

However I'm not sure if it is a good idea to include it into library typings as it's more like a `nice-to-have` thing than a really required part of the typings.

I mean anyone who wants to use it can just copy the definition from your article (it's pretty small), and even if we will ship this definition with `@types/react-redux` I'm pretty much sure that 99% of users will never notice that and will not use it (unless your article will be promoted so much that all of these 99% of users will read it, of course).

To sum it up: I think it will be more beneficial to promote the article itself then including `TypeOfConnect` higher-order-type into library typings; however if there will be a lot of developers who really want to have it included - I don't really mind.

**Comment by charrondev** [comment]
I think it would be beneficial to have it included. Utility types for dealing with the react/redux boundaries are always helpful!

**Comment by markerikson** [comment]
Hi!  I'm a Redux maintainer.  And, uh, I'm finally starting to use Redux with TypeScript for realzies now.

I just ran into this issue myself, and I would _strongly_ recommend that this "extract the connected props" type be included in the library.  If it is, we can then document it as a recommended way to handle things.

**Comment by charrondev** [comment]
I will note that my current workaround tends to look something like this:

```tsx

function MyComponent(props: IProps) {
   return <div /> 
}

interface IOwnProps {
   someParam: boolean;
}

function mapStateToProps(state: IStoreState) {
    return {
        // Return some stuff.
    }
}

function mapDispatchToProps(dispatch: IDispatch) {
    return {
        // Return some stuff.
    }
}

type IProps = IOwnProps & ReturnType<typeof mapStateToProps> & ReturnType<typeof mapDisaptchToProps>;

export default connect(mapStateToProps, mapDispatchToProps)(MyComponent);

```

**Comment by markerikson** [comment]
Yeah, that's basically what I was doing, except that I was using the object shorthand form of `mapDispatch`:

```ts
const mapState = (state: RootState) => ({
    todos: state.todos,
})

const mapDispatch = {
    fetchTodosThunk,
}

interface OwnProps {
    someValue: string;
}

type MyComponentProps = ReturnType<mapState> & typeof mapDispatch & OwnProps;
```

But, this breaks because `connect` "unwraps" the thunk return types, so you get a mismatch: `fetchTodosThunk` is defined as `() => SomeThunkType`, while the unwrapped result is `() => void`.

By reusing the React-Redux internal types, we can change it to:

```ts
const mapState = (state: RootState) => ({
    todos: state.todos,
})

const mapDispatch = {
    fetchTodosThunk,
}

interface OwnProps {
    someValue: string;
}

const connectMyComponent = connect(mapState, mapDispatch);

type PropsFromRedux = ConnectedProps<typeof connectMyComponent>;
type MyComponentProps = PropsFromRedux & OwnProps;

class MyComponent extends Component<MyComponentProps> {}

export default connectMyComponent(MyComponent);
```

Still several steps, but at least it fixes the issues with using `mapDispatch` as an object, and simplifies the "props from Redux" part.

**Comment by markerikson** [comment]
Hi, it's me again!

I would _really_ like to have this added in to the React-Redux types, per https://gist.github.com/christianchown/05e084c78ec216070a5a2b80f0534d4b:

```ts
import { InferableComponentEnhancerWithProps } from 'react-redux';

export default type ConnectedProps<T> = T extends InferableComponentEnhancerWithProps<infer Props, infer _> ? Props : never;
```

Who owns the typedefs, and how do we make this happen?

**Comment by markerikson** [comment]
Waitasec.  The `ConnectedProps` type was already added in #37300 .

This issue can be closed, as the type now exists and was published in https://unpkg.com/browse/@types/react-redux@7.1.2/ .

**Comment by Voronar** [comment]
Thank you for adding this stuff here.
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/37300

**Comment by markerikson** [comment]
The `ConnectedProps<T>` pattern is now described in the React-Redux docs, and we're officially recommending that people use it:

https://react-redux.js.org/using-react-redux/static-typing#typing-the-connect-higher-order-component
